File: starter_kits/Java/hlt/Log.java
package hlt;

import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;

public class Log {
    private final FileWriter file;

    private static Log INSTANCE;
    private static ArrayList<String> LOG_BUFFER = new ArrayList<>();

    static {
        Runtime.getRuntime().addShutdownHook(new AtExit());
    }

    private static class AtExit extends Thread {
        @Override
        public void run() {
            if (INSTANCE != null) {
                return;
            }

            final long now_in_nanos = System.nanoTime();
            final String filename = "bot-unknown-" + now_in_nanos + ".log";
            try (final FileWriter writer = new FileWriter(filename)) {
                for (final String message : LOG_BUFFER) {
                    writer.append(message).append('\n');
                }
            } catch (final IOException e) {
                // Nothing much we can do here.
            }
        }
    }

    private Log(final FileWriter f) {
        file = f;
    }

    static void open(final int botId) {
        if (INSTANCE != null) {
            Log.log("Error: log: tried to open(" + botId + ") but we have already opened before.");
            throw new IllegalStateException();
        }

        final String filename = "bot-" + botId + ".log";
        final FileWriter writer;
        try {
            writer = new FileWriter(filename);
        } catch (final IOException e) {
            throw new IllegalStateException(e);
        }
        INSTANCE = new Log(writer);

        try {
            for (final String message : LOG_BUFFER) {
                writer.append(message).append('\n');
            }
        } catch (final IOException e) {
            throw new IllegalStateException(e);
        }
        LOG_BUFFER.clear();
    }

    public static void log(final String message) {
        if (INSTANCE == null) {
            LOG_BUFFER.add(message);
            return;
        }

        try {
            INSTANCE.file.append(message).append('\n').flush();
        } catch (final IOException e) {
            e.printStackTrace();
        }
    }
}

File: starter_kits/Kotlin/hlt/Log.kt
package hlt

import java.io.FileWriter
import java.io.IOException
import java.util.*

class Log private constructor(private val file: FileWriter) {

    private class AtExit : Thread() {
        override fun run() {
            if (INSTANCE != null) {
                return
            }

            val now_in_nanos = System.nanoTime()
            val filename = "bot-unknown-$now_in_nanos.log"
            try {
                FileWriter(filename).use { writer ->
                    for (message in LOG_BUFFER) {
                        writer.append(message).append('\n')
                    }
                }
            } catch (e: IOException) {
                // Nothing much we can do here.
            }

        }
    }

    companion object {

        private var INSTANCE: Log? = null
        private val LOG_BUFFER = ArrayList<String>()

        init {
            Runtime.getRuntime().addShutdownHook(AtExit())
        }

        internal fun open(botId: Int) {
            if (INSTANCE != null) {
                Log.log("Error: log: tried to open($botId) but we have already opened before.")
                throw IllegalStateException()
            }

            val filename = "bot-$botId.log"
            val writer: FileWriter
            try {
                writer = FileWriter(filename)
            } catch (e: IOException) {
                throw IllegalStateException(e)
            }

            INSTANCE = Log(writer)

            try {
                for (message in LOG_BUFFER) {
                    writer.append(message).append('\n')
                }
            } catch (e: IOException) {
                throw IllegalStateException(e)
            }

            LOG_BUFFER.clear()
        }

        fun log(message: String) {
            if (INSTANCE == null) {
                LOG_BUFFER.add(message)
                return
            }

            try {
                INSTANCE!!.file.append(message).append('\n').flush()
            } catch (e: IOException) {
                e.printStackTrace()
            }

        }
    }
}
